File: leetcode/c++/6.ZigZag-Conversion.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <numeric>      


using namespace std;

string convert(string s,int numRows){
  if(numRows==1||s.length()<numRows){
    return s;
  }

  vector<vector<char>> rows;
  int currentRow=0;
  bool reverse = false;
  string result="";

  for(int i=0;i<numRows;i++){
    vector<char> s;
    rows.push_back(s);
  }

  for(int i=0;i<s.length();i++){
    rows[currentRow].push_back(s[i]);

    if(!reverse){
        currentRow++;
    }else{
        currentRow--;
    }

    if(currentRow == numRows -1 || currentRow ==0 ){
      reverse = !reverse;
    }
  }
  
  for(auto i:rows){
   result+= accumulate(
    i.begin(),i.end(),string("")
  );
  }

  return result;
}

int main() {
  cout << convert("PAYPALISHIRING", 3)<<endl;
  return 0;
}
// PAHNAPLSIIGYIR

File: leetcode/python/6.ZigZag-Conversion.py
def convert(s, numRows):
    if numRows == 1 or len(s) < numRows:
        return s

    rows = []
    currentRow = 0
    reverse = False
    result = ""

    for i in range(0, numRows):
        rows.insert(i, [])

    for i in range(0, len(s)):
        rows[currentRow].append(s[i])

        if reverse == False:
            currentRow += 1
        else:
            currentRow -= 1

        if currentRow == numRows - 1 or currentRow == 0:
            reverse = not reverse

    for i in rows:
        result += "".join(i)

    return result


print(convert("PAYPALISHIRING", 3))
# PAHNAPLSIIGYIR